lib/editors.js

```diff
@@ -389,6 +389,97 @@ class Editor {
         });
     }
 
+    /**
+     * Remove empty preceding subfolders for a question, assessment, etc. based on its ID.
+     * This should be run after renames or deletes to prevent syncing issues.
+     * @param rootDirectory Root directory that the items are being stored in.
+     * @param id Item to delete root subfolders for, relative from the root directory.
+     * @callback {function(err)} Function to call once this has finished.
+     */
+    removeEmptyPrecedingSubfolders(rootDirectory, id, callback) {
+        const idSplit = id.split(path.sep);
+
+        /* Start deleting subfolders in reverse order */
+        const reverseFolders = idSplit.slice(0, -1).reverse();
+        debug('Checking folders', reverseFolders);
+
+        let seenNonemptyFolder = false;
+        async.eachOfSeries(reverseFolders, (folder, index, callback) => {
+            if (!seenNonemptyFolder) {
+                const delPath = path.join(rootDirectory, ...idSplit.slice(0, idSplit.length - 1 - index));
+                debug('Checking', delPath);
+
+                fs.readdir(delPath, (err, files) => {
+                    if (ERR(err, callback)) return;
+
+                    /* Delete the subfolder if it's empty, otherwise stop here */
+                    if (files.length > 0) {
+                        debug(delPath, 'is nonempty, stopping here.');
+                        debug('Folder contains', files);
+                        seenNonemptyFolder = true;
+                        callback(null);
+                    } else {
+                        debug('No files, deleting', delPath);
+                        fs.remove(delPath, (err) => {
+                            if (ERR(err, callback)) return;
+                            callback(null);
+                        });
+                    }
+                });
+            } else { callback(null); }
+        }, callback);
```

Comment:
See the second bullet point here: https://prairielearn.readthedocs.io/en/latest/dev-guide/#stack-traces-with-callback-style-functions